File: app/src/main/java/com/digital/playground/core/BaseObservable.java
package com.digital.playground.core;

import com.digital.playground.util.AppUtils;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.disposables.Disposable;

public abstract class BaseObservable<T> implements Observer<T> {

    @Override
    public void onSubscribe(@NonNull Disposable dispo) {
        /* Disposable */
    }

    @Override
    public void onNext(@NonNull T t) {
        success(t);
    }

    @Override
    public void onError(@NonNull Throwable t) {
        failure(t);
        AppUtils.showException(t);
    }

    @Override
    public void onComplete() {
    }

    public abstract void success(T response);

    public abstract void failure(Throwable t);
}

File: app/src/main/java/com/digital/playground/core/BaseObservable.kt
package com.digital.playground.core

import com.digital.playground.util.AppUtils.showException
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable

abstract class BaseObservable<T> : Observer<T> {
    override fun onSubscribe(dispo: Disposable) {
        /* Disposable */
    }

    override fun onNext(t: T) {
        success(t)
    }

    override fun onError(t: Throwable) {
        failure(t)
        showException(t)
    }

    override fun onComplete() {}
    abstract fun success(response: T)
    abstract fun failure(t: Throwable)
}